contracts/src/v0.1/L2endpoint.sol

```diff
@@ -53,4 +72,109 @@ contract L2Endpoint is Ownable {
+        });
+        emit RandomWordsRequested(
+            keyHash,
+            requestId,
+            preSeed,
+            accId,
+            callbackGasLimit,
+            numWords,
+            msg.sender,
+            isDirectPayment
+        );
+
+        return requestId;
+    }
+
+    /**
+     * @dev calls target address with exactly gasAmount gas and data as calldata
+     * or reverts if at least gasAmount gas is not available.
+     */
+    function callWithExactGas(
+        uint256 gasAmount,
+        address target,
+        bytes memory data
+    ) internal returns (bool success) {
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            let g := gas()
+            // Compute g -= GAS_FOR_CALL_EXACT_CHECK and check for underflow
+            // The gas actually passed to the callee is min(gasAmount, 63//64*gas available).
+            // We want to ensure that we revert if gasAmount >  63//64*gas available
+            // as we do not want to provide them with less, however that check itself costs
+            // gas.  GAS_FOR_CALL_EXACT_CHECK ensures we have at least enough gas to be able
+            // to revert if gasAmount >  63//64*gas available.
+            if lt(g, GAS_FOR_CALL_EXACT_CHECK) {
+                revert(0, 0)
+            }
+            g := sub(g, GAS_FOR_CALL_EXACT_CHECK)
+            // if g - g//64 <= gasAmount, revert
+            // (we subtract g//64 because of EIP-150)
+            if iszero(gt(sub(g, div(g, 64)), gasAmount)) {
+                revert(0, 0)
+            }
+            // solidity calls check that a contract actually exists at the destination, so we do the same
+            if iszero(extcodesize(target)) {
+                revert(0, 0)
+            }
+            // call and return whether we succeeded. ignore return data
+            // call(gas,addr,value,argsOffset,argsLength,retOffset,retLength)
+            success := call(gasAmount, target, 0, add(data, 0x20), mload(data), 0, 0)
+        }
+        return success;
+    }
+
+    function requestRandomWords(
+        bytes32 keyHash,
+        uint64 accId,
+        uint32 callbackGasLimit,
+        uint32 numWords
+    ) external returns (uint256) {
+        bool isDirectPayment = false;
+        uint256 requestId = requestRandomWords(
+            keyHash,
+            accId,
+            callbackGasLimit,
+            numWords,
+            isDirectPayment
+        );
+        return requestId;
+    }
+
+    function fulfillRandomWords(uint256 requestId, uint256[] memory randomWords) external {
+        if (!submitters[msg.sender]) revert InvalidSubmitter(msg.sender);
+        RequestInfo memory detail = sRequestDetail[requestId];
+        bytes memory resp = abi.encodeWithSelector(
+            VRFConsumerBase.rawFulfillRandomWords.selector,
+            requestId,
+            randomWords
+        );
+        bool success = callWithExactGas(detail.callbackGasLimit, detail.owner, resp);
```

Comment:
I think we should apply reentrancy lock for `requestRandomWords` and `fullfillRandomWords` function as in `VRFCoordinator`

https://github.com/Bisonai/orakl/blob/ae1e7efd98d461360e473d939fda589633d262d8/contracts/src/v0.1/CoordinatorBase.sol#L51-L56

https://github.com/Bisonai/orakl/blob/ae1e7efd98d461360e473d939fda589633d262d8/contracts/src/v0.1/VRFCoordinator.sol#L161

https://github.com/Bisonai/orakl/blob/ae1e7efd98d461360e473d939fda589633d262d8/contracts/src/v0.1/VRFCoordinator.sol#L242

and we should include the lock around `callWithExactGas`. 

https://github.com/Bisonai/orakl/blob/ae1e7efd98d461360e473d939fda589633d262d8/contracts/src/v0.1/VRFCoordinator.sol#L266-L268